infra/ballista/build/lambda-handler.py

```diff
@@ -0,0 +1,154 @@
+import base64
+import os
+import subprocess
+from subprocess import PIPE
+import logging
+import pexpect
+from pexpect import popen_spawn
+import time
+
+
+logging.getLogger().setLevel(logging.DEBUG)
+
+# Create global variables for the forked processes
+process_s = None
+process_e = None
+process_cli = None
+IS_COLD_START = True
+
+
+def init():
+    # start scheduler
+    global process_s
+    process_s = subprocess.Popen(
+        ["/opt/ballista/ballista-scheduler"], stdout=PIPE, stderr=PIPE
+    )
+    logging.debug("scheduler starts")
+    # wait till the scheduler is up and running
+    return_code = 1
+    while return_code:
+        res = subprocess.run(["nc", "-z", "localhost", "50050"], capture_output=True)
```

Comment:
❕ Instead of spinning up an `nc` process, you could also:
```python
import socket
from contextlib import closing

while True:
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        if sock.connect_ex(("localhost", 8786)) == 0:
            break
```

Reply:
Interesting, could you elaborate on the benefit of this approach?

Reply:
- one less thing to install in the image 
- it is confusing when you read the dockerfile while `nc` is required
- calling an external process is [expensive](https://cs.stackexchange.com/questions/129410/overhead-cost-of-spawning-child-processes), so it's probably better to avoid it in a polling while loop

Reply:
your right about clarity, I'll give it a try